NXP drivers: counter_mcux_ctimer: Adds PM low-power recovery support

tsi-chung opened the pull request:
Enables Sleep mode (PM3) in RW61x for Ctimer.

tsi-chung committed:
Counter_mcux_ctimer: Adds PM low-power recovery support

Enables Sleep mode (PM3) in RW61x for Ctimer.

Signed-off-by: Tsi-Chung Liew <Tsi-Chung.Liew@nxp.com>

decsny commented on tests/drivers/counter/counter_basic_api/boards/frdm_rw612.conf:
```diff
@@ -0,0 +1,9 @@
+#
+# SPDX-License-Identifier Apache-2.0
+#
+CONFIG_COUNTER_MCUX_CTIMER=y
```
shouldn't be needed, it gets enabled if DT node is enabled

decsny commented on drivers/counter/counter_mcux_ctimer.c:
```diff
@@ -277,6 +279,32 @@ static int mcux_lpc_ctimer_init(const struct device *dev)
+
+static int mcux_lpc_ctimer_init(const struct device *dev)
+{
+	/* Rest of the init is done from the PM_DEVICE_TURN_ON action
+	 * which is invoked by pm_device_driver_init().
```
you can't do this paradigm because PM device could (and most likely will) be off on most platforms and app configurations

mmahadevan108 replied:
@decsny . This implementation was added per this comment.
https://github.com/zephyrproject-rtos/zephyr/pull/84945#discussion_r1957080778

decsny replied:
ok, I did not realize that having `CONFIG_PM_DEVICE=n` will still build all the PM device files

decsny reviewed (dismissed):
please ensure driver will still work if PM_DEVICE is not configured

mmahadevan108 commented:
@tsi-chung , this PR should address some of the counter test failures you were seeing. https://github.com/zephyrproject-rtos/zephyr/pull/89340

tsi-chung committed:
Counter_nxp_mrt: Adds PM low-power recovery support

Enables Sleep mode (PM3) in RW61x for MRT.

Signed-off-by: Tsi-Chung Liew <Tsi-Chung.Liew@nxp.com>

Commit:
tests: drivers: counter_basic_api: add testcase for NXP FRDM-RW612

New test case enables low-power standby mode, but disables testing
the lpc_rtc_highres driver, since it is used to wake the device from
standby mode.

Signed-off-by: Derek Snell <derek.snell@nxp.com>

decsny commented on drivers/counter/counter_mcux_ctimer.c:
```diff
@@ -11,6 +11,8 @@
 #include <zephyr/drivers/clock_control.h>
 #include <zephyr/dt-bindings/clock/mcux_lpc_syscon_clock.h>
 #include <zephyr/irq.h>
+#include <zephyr/pm/policy.h>
```
I don't see anything from this header being  used in this file

tsi-chung replied:
updated. Will push later

decsny commented on drivers/counter/counter_nxp_mrt.c:
```diff
@@ -22,6 +22,8 @@
 #include <zephyr/device.h>
 #include <zephyr/irq.h>
 #include <zephyr/drivers/reset.h>
+#include <zephyr/pm/policy.h>
```
I don't see anything from this header being  used in this file

tsi-chung replied:
updated. Will push later

tsi-chung commented on tests/drivers/counter/counter_basic_api/boards/frdm_rw612.conf:
```diff
@@ -0,0 +1,10 @@
+CONFIG_COUNTER_NXP_MRT=n
+
+CONFIG_PM=y
+CONFIG_PM_DEVICE=y
```
this link https://docs.zephyrproject.org/latest/services/pm/device.html said that set CONFIG_PM_DEVICE; however, there isn't any doc said that not to set it.

tsi-chung commented on tests/drivers/counter/counter_basic_api/boards/frdm_rw612.conf:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier Apache-2.0
+#
+CONFIG_COUNTER_MCUX_CTIMER=n
+CONFIG_COUNTER_NXP_MRT=n
```
updated and pushed
